sheet_to_triples/tests/test_main.py

```diff
@@ -0,0 +1,267 @@
+
+
+def _mock_open(data):
+    return mock.patch('builtins.open', mock.mock_open(read_data=data))
+
+
+def _mock_stderr(buffer):
+    return mock.patch('sys.stderr', new=buffer)
+
+
+_iter_from_name = 'sheet_to_triples.trans.Transform'
+
+
+@mock.patch(_iter_from_name, new=StubTransform('dummy'))
+class TestParseArgs(unittest.TestCase):
+
+    def test_parse_args_default(self):
+        argv = ['dummy', 'transform1']
+        args = main.parse_args(argv)
+        defaults = [
+            ('book', None),
+            ('add_graph', None),
+            ('model', None),
+            ('model_out', 'new.json'),
+            ('resolve_same', True),
+            ('debug', False),
+            ('verbose', False),
+            ('from_list', None),
+            ('non_unique_from', None),
+        ]
+
+        for arg, expected in defaults:
+            with self.subTest(arg=arg):
+                self.assertEqual(getattr(args, arg), expected)
+
+    def test_parse_args_transform(self):
+        argv = ['dummy', 'transform1', 'transform2']
+        args = main.parse_args(argv)
+
+        self.assertEqual(
+            [t.name for t in args.transform],
+            ['transform1.py', 'transform2.py']
+        )
+
+    def test_parse_args_transform_error_if_requires_book(self):
+        argv = ['dummy', 'booktransform1']
+        buffer = io.StringIO()
+
+        # this makes any unexpected lower level exceptions harder to debug
+        # TODO: come up with a nicer way of suppressing "normal" error output
+        with _mock_stderr(buffer), self.assertRaises(SystemExit) as e:
+            main.parse_args(argv)
+
+        self.assertEqual(e.exception.code, 2)
+
+        # argparse spits out usage info so need to isolate error message
+        message = buffer.getvalue().strip().splitlines()[-1]
+        self.assertEqual(
+            message,
+            'dummy: error: transforms {\'booktransform1.py\'} require --book'
+        )
+
+    def test_parse_args_book_is_book(self):
+        argv = [
+            'dummy', 'transform1',
+            '--book', 'book.xlsx',
+            '--book', 'book.xls',
+        ]
+
+        with _mock_os_path_isdir(False) as isdir:
+            args = main.parse_args(argv)
+
+        expected = ['book.xlsx', 'book.xls']
+        self.assertEqual([b for b in args.book], expected)
+
+        self.assertEqual(
+            [c.args[0] for c in isdir.call_args_list],
+            expected
+        )
+
+    def test_parse_args_book_is_dir(self):
+        argv = [
+            'dummy', 'transform1',
+            '--book', 'rootdir1'
+        ]
+        walk_data = [
+            ('rootdir1', '_', ('book.xlsx', 'notabook', 'book.xls')),
+        ]
+
+        with _mock_os_path_isdir(True) as isdir, \
+                _mock_os_path_walk(walk_data):
+            args = main.parse_args(argv)
+
+        expected = ['rootdir1/book.xlsx', 'rootdir1/book.xls']
+        self.assertEqual([b for b in args.book], expected)
```

Comment:
This fails on windows? Haven't got a great way of avoiding this fs logic inline, but [`mockfs`](https://mockfs.github.io/) looks pretty reasonable.

Reply:
There's also [`pyfakefs`](http://jmcgeheeiv.github.io/pyfakefs/release/index.html) which is the same basic idea.